src/Compilers/CSharp/Portable/Binder/Binder_Conversions.cs

```diff
@@ -128,6 +133,88 @@ internal partial class Binder
+            BindArgumentsAndNames(node.ArgumentList, diagnostics, arguments, allowArglist: true);
+
+            BoundObjectInitializerExpressionBase boundInitializerOpt = node.Initializer != null
+                ? BindInitializerExpression(syntax: node.Initializer, type: destination, typeSyntax: syntax, diagnostics)
+                : null;
+
+            BoundExpression boundCreation = null;
+
+            switch (destination.TypeKind)
+            {
+                case TypeKind.Array:
+                case TypeKind.Enum:
+                case TypeKind.Delegate:
+                case TypeKind.Interface:
+                    Error(diagnostics, ErrorCode.ERR_BadTargetTypeForNew, syntax, destination);
+                    break;
+
+                case TypeKind.Pointer:
+                    Error(diagnostics, ErrorCode.ERR_UnsafeTypeInObjectCreation, syntax, destination);
+                    break;
+
+                case TypeKind.Dynamic:
+                    Error(diagnostics, ErrorCode.ERR_NoConstructors, syntax, destination);
+                    break;
+
+                case TypeKind.Struct:
+                    if (destination.IsTupleType)
+                    {
+                        Error(diagnostics, ErrorCode.ERR_NewWithTupleTypeSyntax, syntax, destination);
+                        break;
+                    }
+
+                    goto case TypeKind.Class;
+
+                case TypeKind.Class:
+                    boundCreation = BindClassCreationExpression(
+                        node,
+                        typeName: destination.Name,
+                        typeNode: node,
+                        type: (NamedTypeSymbol)destination,
+                        arguments,
+                        diagnostics,
+                        boundInitializerOpt,
+                        forTargetTypedNew: true);
+                    break;
+
+                case TypeKind.TypeParameter:
+                    var typeParameter = (TypeParameterSymbol)destination;
+                    if (!typeParameter.IsInstantiable())
+                    {
+                        Error(diagnostics, ErrorCode.ERR_BadTargetTypeForNew, syntax, destination);
+                        break;
+                    }
+
+                    boundCreation = BindTypeParameterCreationExpression(
+                       node,
+                       typeParameter,
+                       boundInitializerOpt,
+                       diagnostics);
+                    break;
+            }
+
+            arguments.Free();
+
+            return new BoundConversion(
+                syntax,
+                operand: boundCreation ?? unboundCreation,
+                conversion,
+                @checked: false,
+                explicitCastInCode: isCast,
+                constantValueOpt: null,
+                type: destination,
+                hasErrors: boundCreation is null)
+            { WasCompilerGenerated = source.WasCompilerGenerated };
```

Comment:
>WasCompilerGenerated [](start = 14, length = 20)

Feels like this should be `WasCompilerGenerated = true`, since the "implicit new" conversion can never be explicit in source.

Reply:
Au contraire.  The "implicit new" conversion is explicit in the following source:
``` c#
    var x = (int)(new());
```

A conversion cannot both be compiler-generated and `explicitCastInCode`. This translation currently produces no conversion for this program that is not considered compiler-generated. It must.

---
In reply to: [202206149](https://github.com/dotnet/roslyn/pull/25196#discussion_r202206149) [](ancestors = 202206149)